### PR TITLE
Fix and rewrite markup anchor processing

### LINKED ISSUE #29877: Broken links to anchors for non-ASCII symbols on wiki TOC
### Description

Anchors with non-ASCII symbols encodes twice and then isn't working anymore. Disabling javascript helps.

Create wiki page with title
```
# 气候宜人
```
It renders as `#user-content-%E6%B0%94%E5%80%99%E5%AE%9C%E4%BA%BA` but javascript modifies this to `#%25E6%25B0%2594%25E5%2580%2599%25E5%25AE%259C%25E4%25BA%25BA`

Seems it is regression since #29666 #29305

### Gitea Version

1.21.8

### Can you reproduce the bug on the Gitea demo site?

Yes

### Log Gist

_No response_

### Screenshots

_No response_

### Git Version

_No response_

### Operating System

_No response_

### How are you running Gitea?

from command line

### Database

SQLite

### PR DESCRIPTION
**Pull request by lunny**
Fix #29877

### CODE REVIEW

**Comment by wxiaoguang** [comment]
Why sometimes encode but sometimes not?


image

**Review by delvh — APPROVED** [review]

**Review by lafriks — APPROVED** [review]

**Comment by lunny** [comment]
> Why sometimes encode but sometimes not?
> 
> image

014c3a2

**Comment by silverwind** [comment]
Not sure this is right. I'd certainly want to add some tests for sure.

**Comment by wxiaoguang** [comment]
I don't know whether it is right either .... does the author or approvers have more information to write into comment?

**Review by silverwind — CHANGES_REQUESTED** [review]
needs deeper investigation, i will do later.

**Comment by silverwind** [comment]
As per https://stackoverflow.com/a/50141030/808699 and https://jsfiddle.net/silverwind/pe0twrf8/, it seems unnecssary to encode when setting a `href` attribute.

**Comment by silverwind** [comment]
Fiddle updated: https://jsfiddle.net/silverwind/pe0twrf8/2/

My conclusion is don't encode when writing into DOM, but do decode when reading from DOM.

File: web_src/js/markup/anchors.js [hunk 1]
```diff
@@ -25,7 +25,7 @@ export function initMarkupAnchors() {
     const originalId = heading.id.replace(/^user-content-/, '');
     const a = document.createElement('a');
     a.classList.add('anchor');
-    a.setAttribute('href', `#${encodeURIComponent(originalId)}`);
+    a.setAttribute('href', `#${decodeURIComponent(originalId)}`);
```
Comment (silverwind) [inline web_src/js/markup/anchors.js]:
```suggestion
    a.setAttribute('href', `#${originalId}`);
```

File: web_src/js/markup/anchors.js [hunk 2]
```diff
@@ -38,8 +38,8 @@ export function initMarkupAnchors() {
     const href = a.getAttribute('href');
     if (!href.startsWith('#user-content-')) continue;
     const originalId = href.replace(/^#user-content-/, '');
-    a.setAttribute('href', `#${encodeURIComponent(originalId)}`);
-    if (a.closest('.markup').querySelectorAll(`a[name="${originalId}"]`).length !== 1) {
+    a.setAttribute('href', `#${decodeURIComponent(originalId)}`);
+    if (a.closest('.markup').querySelectorAll(`a[name="${decodeURIComponent(originalId)}"]`).length !== 1) {
```
Comment (silverwind) [inline web_src/js/markup/anchors.js]:
```suggestion
    a.setAttribute('href', `#${originalId}`);
    if (a.closest('.markup').querySelectorAll(`a[name="${CSS.escape(originalId)}"]`).length !== 1) {
```

**Comment by silverwind** [comment]
Apparently it does make a difference whether a link is generated or pre-existing in HTML:

https://jsfiddle.net/silverwind/2wy39d7e/

So backend-generated links will appear escaped and need a `decodeURIComponent` when reading them from DOM. JS-generated links can insert as unicode into DOM, but I guess for consistency we should also encode them.

**Comment by silverwind** [comment]
I rewrote this file quite a bit now, the code is now easier to read and there is no more complex `querySelector` and only a single `click` listener that does it all.

Everything is working and I specifically tested with chinese links as well. @lunny can you try you test case on wiki as well?

**Review by silverwind — APPROVED** [review]

**Comment by silverwind** [comment]
I've invalidated existing reviews because of substantial changes.

File: web_src/js/markup/anchors.js [hunk 3]
```diff
@@ -1,50 +1,67 @@
...
+      const originalId = href.replace(/^#user-content-/, '');
+      a.setAttribute('href', `#${originalId}`);
+    }
+
+    // add `user-content-` prefix to user-generated `a[name]` link targets
+    for (const a of markupEl.querySelectorAll('a[name]')) {
+      const name = a.getAttribute('name');
+      if (!name) continue;
+      a.setAttribute('name', `user-content-${a.name}`);
+    }
```
Comment (silverwind) [inline web_src/js/markup/anchors.js:58]: This prefix addition is matching GitHub behaviour. Ideally we would do this prefixing in backend later, but I didn't want to touch backend here.

**Review by lafriks — APPROVED** [review]

**Comment by lunny** [comment]
It works but the links are still encoded which is different from Github's implementations. For modern web browsers, we don't need to encode them.

image

**Comment by silverwind** [comment]
> It works but the links are still encoded which is different from Github's implementations. For modern web browsers, we don't need to encode them.

Yes, but it comes out encoded from the backend. I assume from the markup renderer, so would need to be fixed there first. If you can, raise a PR and then I will adapt the JS in the same PR to support this.

**Pull request merged**